### ISSUE TITLE
Alerting

### ISSUE THREAD

**Issue by Vadorequest** [FidelLimited/serverless-plugin-warmup#179]
I wonder how alerts (health checks) work with this plugin.

It's meant to be a warmup tool, but I don't see the point of using both warmup and heath check as different tools. Since the lambda is hit during the warmup, it's the perfect time to send an alert if something isn't right.

I couldn't find any option regarding alerts, but I may have missed something.

My need is that I currently use AWS HealthChecks to both warmup and perform heath checks. But I don't like it for the following reasons:

- It's not really flexible. HealthCheck forces at least to send calls from 3 A-Z and the maximum delay is 30 sec. Those two limitations lead to an average of 1 call every 10 secs, at least.
- It's costly. (not so much, but there are many hidden option at $1 kinda meant to be enabled by mistake)
- I didn't succeed to automatically configure alerts using the serverless framework, I had to configure them manually, and that doesn't work anymore from the aws console for some reason

In short, I'm looking for a replacement. And this plugin seems to be the closest to what I seek. 

But I am not sure about what happens if a lambda returns a 5xx error, can we be warned somehow? The goal is to eventually get an alert in a slack channel stating that the service is down.

**Comment by juanjoDiaz** [comment]
Hi @Vadorequest ,

Do you mean ELB HealthChecks or what kind of HealthChecks?

This plugin just logs the responses from the lambdas to CloudWatch (see https://github.com/FidelLimited/serverless-plugin-warmup/blob/master/src/index.js#L346).
You can set up a CloudWatch alarm for errors if you want. you can use the serverless resource section or even https://github.com/ACloudGuru/serverless-plugin-aws-alerts.

I'm open for suggestions on how to improve this 🙂

**Comment by Vadorequest** [comment]
Hi @juanjoDiaz  :) 

I meant Route53 HealthChecks, I wasn't aware there was multiple kind of Health checks on AWS 😅 

https://console.aws.amazon.com/route53/healthchecks/home
![image](https://user-images.githubusercontent.com/3807458/64596303-ef5d8100-d3b3-11e9-84c5-6c8fe24ed266.png)

Is there any example using https://github.com/ACloudGuru/serverless-plugin-aws-alerts by any chance?

**Comment by Vadorequest** [comment]
Regarding pricing, I just compared AWS Route53 HealthCheck vs AWS Lambda:
- https://aws.amazon.com/route53/pricing/?nc1=h_ls
- https://aws.amazon.com/lambda/pricing/?nc1=h_ls

**Route 53:**
- $0.50* per health check / month
- $1.00 / month per optional feature:
  - HTTPS
  - String Matching
  - Fast Interval
  - Latency Measurement
> Monthly charge: $1.50 (with HTTPS option)

**Lambda:** (using 128Mo RAM)
- 1 million free/month
- $0.0000002 per request
- $0.000000208 per 100ms

> Monthly compute charges (monthly): 0.51 USD (no free tiers, with 1 250 000 requests a month)

![image](https://user-images.githubusercontent.com/3807458/64597036-52034c80-d3b5-11e9-8c01-cb64995f8cc6.png)

> I simulated with `1 250 000` requests/months because that's how many requests Route53 HealthChecks send **at least** a month (super optimised), with only 3 datacenters and 30 sec delay

---
Simulation indicates that using Lambda instead of AWS Route53 HealthChecks cost around 3 times less

**Comment by juanjoDiaz** [comment]
Hi @Vadorequest ,

Sorry for the delay responding.
Could you clarify what your question or request is?
I'm not very clear at the moment 🙂

**Comment by Vadorequest** [comment]
Hi @juanjoDiaz :)

Currently, I'm using AWS Route53 HeathChecks to perform warmups (I don't use this plugin yet), but as it lacks configuration capabilities, I'm looking for something else (and cheaper, if possible). 

This plugin seems to do about everything I need, but I wonder how I can get alerts in my Slack channel when my endpoints are unreachable.

The goal of this plugin is to perform warmups. But I believe it would make sense to also perform health checks at the same time. The goal is different, but the mean is very similar: Calling a given endpoint at a regular interval.

I haven't looked into it deeply, but maybe https://github.com/ACloudGuru/serverless-plugin-aws-alerts could help on the "alerting" part, as you stated before.

**Comment by juanjoDiaz** [comment]
Hi @Vadorequest 

I understand now what you mean.
However, warming up and health checking is not the same, especially if using concurrency.

In general, I'd argue that you don't need to health check lambdas since AWS will take care of removing faulty lambdas and replace them with fresh containers.

What you should definitely do is monitor if your lambdas are erroring or taking longer than expected. For that AWS CloudWatch and https://github.com/ACloudGuru/serverless-plugin-aws-alerts are definitely the way to go in my opinion. Of course, you can also use NewRelic, DataDog or any other APM solution.

In that sense, this plugin logs to the console whether the calls to the lambdas being warmed up succeded or not. So you could set alarms based on those logs. Wouldn't that solve your use case?
If you think that the logging of this plugin could be improved we can also work on that. 🙂

**Comment by Vadorequest** [comment]
Thanks for your advices @juanjoDiaz, much welcome.

I think I need to take a deeper look at https://github.com/ACloudGuru/serverless-plugin-aws-alerts and see how it goes from there. :)

**Comment by juanjoDiaz** [comment]
No problem!

I'll close this issue.
Feel free to reopen or open a new issue if after your research you find improvements or changes needed in this plugin.
